pkg/runtime/runtime.go

```diff
@@ -2138,13 +2138,27 @@ func (a *DaprRuntime) BulkPublish(req *pubsub.BulkPublishRequest) (pubsub.BulkPu
 	if allowed := a.isPubSubOperationAllowed(req.PubsubName, req.Topic, ps.ScopedPublishings); !allowed {
 		return pubsub.BulkPublishResponse{}, runtimePubsub.NotAllowedError{Topic: req.Topic, ID: a.runtimeConfig.ID}
 	}
+
 	policyDef := a.resiliency.ComponentOutboundPolicy(req.PubsubName, resiliency.Pubsub)
-	if bulkPublisher, ok := ps.Component.(pubsub.BulkPublisher); ok {
+	bulkPublisher, ok := ps.Component.(pubsub.BulkPublisher)
+
+	if !ok {
+		log.Debugf("pubsub %s does not implement the BulkPublish API; falling back to publishing messages individually", req.PubsubName)
+		defaultBulkPublisher := runtimePubsub.NewDefaultBulkPublisher(ps.Component)
+		return runtimePubsub.ApplyBulkPublishResiliency(context.TODO(), req, policyDef, defaultBulkPublisher)
+	}
+
+	useBulkDefaultPublisher, err := bulkPublisher.UseBulkDefaultPublisher(a.ctx)
```

Comment:
This makes a 2nd network call in each call which is quite a bit of overhead.

Instead of implementing it this way, which also requires changing all components in components-contrib, how about:

1. We implement a new method in the proto for pubsub pluggable components called "GetMetadata" which returns metadata about the component, including capabilities
2. We list if the component supports native bulk publish
3. The GetMetadata method is invoked just once when the component is initialized. Its status is kept in the component registry, since it's not going to change in the future.

Reply:
I agree. There should be another way to tag the pubsub as a bulkPubsub without relying on an API call.